e2e_tests/tests/cluster/test_master_restart.py

```diff
@@ -112,3 +137,145 @@ def test_master_restart_kill_works(managed_cluster: ManagedCluster) -> None:
     except Exception:
         managed_cluster.restart_master()
         managed_cluster.restart_agent()
+
+
+@pytest.mark.managed_devcluster
+@pytest.mark.parametrize("slots", [0, 1])
+@pytest.mark.parametrize("downtime", [0, 20, 60])
+def test_master_restart_cmd(
+    restartable_managed_cluster: ManagedCluster, slots: int, downtime: int
+) -> None:
+    managed_cluster = restartable_managed_cluster
+
+    command_id = run_command(30, slots=slots)
+    wait_for_command_state(command_id, "RUNNING", 10)
+
+    if downtime >= 0:
+        managed_cluster.kill_master()
+        time.sleep(downtime)
+        managed_cluster.restart_master()
+
+    wait_for_command_state(command_id, "TERMINATED", 30)
+    succeeded = "success" in get_command_info(command_id)["exitStatus"]
+    assert succeeded
+
+
+@pytest.mark.managed_devcluster
+@pytest.mark.parametrize("downtime", [5])
+def test_master_restart_shell(restartable_managed_cluster: ManagedCluster, downtime: int) -> None:
+    managed_cluster = restartable_managed_cluster
+
+    with cmd.interactive_command("shell", "start", "--detach") as shell:
+        task_id = shell.task_id
+
+        assert task_id is not None
+        wait_for_task_state("shell", task_id, "RUNNING")
+
+        if downtime >= 0:
+            managed_cluster.kill_master()
+            time.sleep(downtime)
+            managed_cluster.restart_master()
+
+        wait_for_task_state("shell", task_id, "RUNNING")
+
+        child = det_spawn(["shell", "open", task_id])
+        child.setecho(True)
+        child.expect(r".*Permanently added.+([0-9a-f-]{36}).+known hosts\.")
+        child.sendline("det user whoami")
+        child.expect("You are logged in as user \\'(.*)\\'", timeout=10)
+        child.sendline("exit")
+        child.read()
+        child.wait()
+        assert child.exitstatus == 0
+
+
+def _get_auth_token_for_curl() -> str:
+    token = authentication.TokenStore(conf.make_master_url()).get_token(
+        constants.DEFAULT_DETERMINED_USER
+    )
+    assert token is not None
+    return token
+
+
+def _check_web_url(url: str, name: str) -> None:
+    token = _get_auth_token_for_curl()
+    bad_status_codes = []
+    for _ in range(10):
+        r = requests.get(url, headers={"Authorization": f"Bearer {token}"}, allow_redirects=True)
+        # Sometimes the TB/JL take a bit of time to stand up, returning 502.
+        # Sometimes it takes a bit of time for master to register the proxy, returning 404.
+        if r.status_code == 502 or r.status_code == 404:
+            time.sleep(1)
+            bad_status_codes.append(r.status_code)
+            continue
+        r.raise_for_status()
+        html = r.content.decode("utf-8")
+        assert name in html  # Brutal.
```

Comment:
lol. for what it is worth, the way the idle check works may be better, at least for notebooks.

Reply:
looks too complicated for a test